src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -49,14 +51,31 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
       this.nodes.clear();
       this.slots.clear();
 
-      String localNodes = jedis.clusterNodes();
-      for (String nodeInfo : localNodes.split("\n")) {
-        ClusterNodeInformation clusterNodeInfo = nodeInfoParser.parse(nodeInfo, new HostAndPort(
-            jedis.getClient().getHost(), jedis.getClient().getPort()));
+      List<Object> slots = jedis.clusterSlots();
 
-        HostAndPort targetNode = clusterNodeInfo.getNode();
-        setNodeIfNotExist(targetNode);
-        assignSlotsToNode(clusterNodeInfo.getAvailableSlots(), targetNode);
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        if (slotInfo.size() <= 2) {
+          continue;
+        }
+
+        List<Integer> slotNums = getAssignedSlotArray(slotInfo);
+
+        // hostInfos
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
```

Comment:
@HeartSaVioR you're right. It's an extreme situation where all the masters should die at exactly the same time, so Jedis won't be able to connect to any. It's __very__ unlikely to happen but not impossible :wink:. We can leave it as it is then.